src/puya/puya-service.ts

```diff
@@ -0,0 +1,166 @@
+import { spawn } from 'child_process'
+import * as rpc from 'vscode-jsonrpc/node'
+import type { SourceLocation } from '../awst/source-location'
+import { logger } from '../logger'
+import { resolvePuyaPath } from './resolve-puya-path'
+
+// Service configuration
+const SERVICE_TIMEOUT_MS = 30000
+const SERVICE_THREADS = 2
+
+// Log levels
+const LOG_LEVELS = {
+  CRITICAL: 50,
+  ERROR: 40,
+  WARNING: 30,
+  INFO: 20,
+  DEBUG: 10,
+  NOTSET: 0,
+} as const
+
+interface PuyaServiceLog {
+  level: string
+  message: string
+  location?: SourceLocation
+}
+
+export type DebugCallback = (message: string) => void
+
+/**
+ * Maps numeric log levels to string literals
+ */
+function mapLogLevel(level: number): string {
+  switch (level) {
+    case LOG_LEVELS.CRITICAL:
+      return 'critical'
+    case LOG_LEVELS.ERROR:
+      return 'error'
+    case LOG_LEVELS.WARNING:
+      return 'warning'
+    case LOG_LEVELS.INFO:
+      return 'info'
+    case LOG_LEVELS.DEBUG:
+      return 'debug'
+    default:
+      return 'notset'
+  }
+}
+
+interface AnalyseParams {
+  awst: unknown
+  compilation_set: Record<string, string>
+  source_annotations?: unknown
+}
+
+interface AnalyseResult {
+  logs: Array<PuyaServiceLog & { level: number }>
+  cancelled: boolean
+}
+
+export class PuyaService {
+  private process: ReturnType<typeof spawn> | null = null
+  private connection: rpc.MessageConnection | null = null
+  private analyseRequest = new rpc.RequestType<AnalyseParams, AnalyseResult, void>('analyse')
+
+  constructor(
+    private cwd: string,
+    private debugCallback?: DebugCallback,
+  ) {}
+
+  setDebugCallback(callback: DebugCallback) {
```

Comment:
What's the debug callback for? Doesn't appear to be used.

Reply:
Removed, this indeed wasn't needed